staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go

```diff
@@ -62,24 +62,57 @@ func (r *BucketRateLimiter) NumRequeues(item interface{}) int {
 func (r *BucketRateLimiter) Forget(item interface{}) {
 }
 
+// itemState holds state for a single item in ItemBucketRateLimiter.
+type itemState struct {
+	// nextBatchStart is a time when the next batch starts.
+	// All requests before this batch should join this batch.
+	nextBatchStart time.Time
+	// limiter is used to rate limit starting new batches.
+	limiter *rate.Limiter
+}
+
 // ItemBucketRateLimiter implements a workqueue ratelimiter API using standard rate.Limiter.
-// Each key is using a separate limiter.
+// The rate limiter connects multiple items with the same key into batches.
+// It always keeps up to one future batch. Items can join that batch as long as it's not started yet.
+// Creating a new batch is rate limited according to qps and burst.
+//
+// Unlike other RateLimiters in this file, this RateLimiter isn't intended to be used to rate
+// limit error retries, rather the successful path.
+//
+// Intended usage:
+// // With qps = 1.0, burst = 1, the first batch can start instantanously, the next every 1s.
+// r := NewItemBucketRateLimiter(1.0, 1)
+//
+// t := time.Now()
+// queue.AddAfter("key1", r.When("key1")) // = AddAfter("key1", 0) => burst = 1 so it can go now.
+// time.Sleep(1 * time.Millisecond)
+//
+// queue.AddAfter("key1", r.When("key1")) // = AddAfter("key1", ~1s) => scheduled to be run at t + 1s
+// queue.AddAfter("key1", r.When("key1")) // = AddAfter("key1", ~1s) => scheduled to be run at t + 1s
+//
+// time.Sleep(1 * time.Second) // The t + 1s batch has started already.
+// queue.AddAfter("key1", r.When("key1")) // = AddAfter("key1", 1s)  // scheduled to be run at the next batch
+// queue.AddAfter("key2", r.When("key2"))
+//
+// // When key1 is deleted from the system.
+// queue.Forget("key1")
+//
 type ItemBucketRateLimiter struct {
-	r     rate.Limit
+	qps   float64
 	burst int
 
 	limitersLock sync.Mutex
-	limiters     map[interface{}]*rate.Limiter
+	limiters     map[interface{}]*itemState
 }
 
 var _ RateLimiter = &ItemBucketRateLimiter{}
 
 // NewItemBucketRateLimiter creates new ItemBucketRateLimiter instance.
-func NewItemBucketRateLimiter(r rate.Limit, burst int) *ItemBucketRateLimiter {
+func NewItemBucketRateLimiter(qps float64, burst int) *ItemBucketRateLimiter {
 	return &ItemBucketRateLimiter{
-		r:        r,
+		qps:      qps,
 		burst:    burst,
-		limiters: make(map[interface{}]*rate.Limiter),
+		limiters: make(map[interface{}]*itemState),
```

Comment:
What keeps this from growing without bound?

Reply:
That's a really good point. It definitely needs to be addressed

/lgtm cancel

Reply:
The user of this class is supposed to call Forget() when given key no longer exists in the system.

Forget() removes objects from this map.

In this case, we should keep only a single itemState object for only for all keys (services) that still exists in the system.